### PR TITLE
v2http: use guest access in non-TLS mode

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/6075.

/cc @xiang90 @shyamradhakrishnan

### CODE REVIEW

File: etcdserver/api/v2http/client_auth.go [hunk 1]
```diff
@@ -116,17 +116,21 @@ func hasKeyPrefixAccess(sec auth.Store, r *http.Request, key string, recursive,
...
-			return false
+	if r.Header.Get("Authorization") == "" {
+		if clientCertAuthEnabled {
+			user = userFromClientCertificate(sec, r)
+			if user == nil {
+				plog.Warningf("auth: no authorization provided, checking guest access")
+				return hasGuestAccess(sec, r, key)
+			}
 		}
+		plog.Warningf("auth: no authorization provided, checking guest access")
```
Comment (xiang90) [inline etcdserver/api/v2http/client_auth.go]: remove this logging.

File: etcdserver/api/v2http/client_auth.go [hunk 2]
```diff
@@ -116,17 +116,21 @@ func hasKeyPrefixAccess(sec auth.Store, r *http.Request, key string, recursive,
...
-		}
-	} else {
-		user = userFromBasicAuth(sec, r)
-		if user == nil {
-			return false
+	if r.Header.Get("Authorization") == "" {
+		if clientCertAuthEnabled {
+			user = userFromClientCertificate(sec, r)
+			if user == nil {
+				plog.Warningf("auth: no authorization provided, checking guest access")
```
Comment (xiang90) [inline etcdserver/api/v2http/client_auth.go]: remove this logging.

File: etcdserver/api/v2http/client_auth.go [hunk 3]
```diff
@@ -116,17 +116,21 @@ func hasKeyPrefixAccess(sec auth.Store, r *http.Request, key string, recursive,
...
-			return hasGuestAccess(sec, r, key)
-		}
-	} else {
-		user = userFromBasicAuth(sec, r)
-		if user == nil {
-			return false
+	if r.Header.Get("Authorization") == "" {
+		if clientCertAuthEnabled {
+			user = userFromClientCertificate(sec, r)
+			if user == nil {
```
Comment (xiang90) [inline etcdserver/api/v2http/client_auth.go]: move this out of the if statement?

```go
if clientCertAuthEnabled {
    user = ...
}

if user == nil {
    return hasGuestAccess(sec, r, key)
}

...
```

File: e2e/ctl_v2_test.go [hunk 4]
```diff
@@ -312,6 +312,30 @@ func TestCtlV2AuthWithCommonName(t *testing.T) {
 	}
 }
 
+func TestCtlV2AuthWithGuestRole(t *testing.T) {
```
Comment (xiang90) [inline e2e/ctl_v2_test.go]: I do not think this is the correct test.

We should set up a cluster with tls + client Auth checking. Then a client with cert that has no valid CN name should be recognized as guest.

**Comment by gyuho** [comment]
@xiang90 All addressed with more comments to the test.

PTAL.

Thanks.

**Comment by gyuho** [comment]
@xiang90 Just added a test case to `TestPrefixAccess` to test `hasKeyPrefixAccess`. I think this is enough to test if empty auth header without client cert enabled correctly uses guess access. If we want the invalid common name test case, we need e2e test to enable client cert auth in etcd binary, I think.

PTAL.

**Comment by xiang90** [comment]
lgtm

**Pull request merged**